### PR TITLE
Rocky Linux 9.4 support

### PR DESCRIPTION
**Pull request by grzegorzkoper**

### CODE REVIEW

File: etc/kayobe/kolla-image-tags.yml [hunk 1]
```diff
@@ -4,25 +4,5 @@
 # where the key is the OS distro and the value is the tag to deploy.
 kolla_image_tags:
   openstack:
-    rocky-9: 2023.1-rocky-9-20240621T104542
+    rocky-9: 2023.1-rocky-9-20240709T132012
     ubuntu-jammy: 2023.1-ubuntu-jammy-20240621T104542
-  bifrost_deploy:
-    rocky-9: 2023.1-rocky-9-20240423T125905
-    ubuntu-jammy: 2023.1-ubuntu-jammy-20240423T125905
```
Comment (mnasiadka) [inline etc/kayobe/kolla-image-tags.yml:11]: Are you sure this should be deleted?

Comment (grzegorzkoper) [reply etc/kayobe/kolla-image-tags.yml:11]: ubuntu-jammy should not . 
Thanks for pointing that out.

**Review by Alex-Welsh — DISMISSED** [review]
LGTM as long as CI passes

File: etc/kayobe/pulp-repo-versions.yml [hunk 2]
```diff
@@ -26,12 +26,17 @@ stackhpc_pulp_repo_rocky_9_2_baseos_version: 20231211T120328
...
+stackhpc_pulp_repo_rocky_9_4_crb_version: 20240702T000233
+stackhpc_pulp_repo_rocky_9_4_extras_version: 20240707T235817
+stackhpc_pulp_repo_rocky_9_4_highavailability_version: 20240629T235004
 stackhpc_pulp_repo_rocky_9_sig_security_common_version: 20240708T235303
-stackhpc_pulp_repo_ubuntu_cloud_archive_version: 20240418T070026
-stackhpc_pulp_repo_ubuntu_jammy_security_version: 20240418T043733
-stackhpc_pulp_repo_ubuntu_jammy_version: 20240418T043733
+stackhpc_pulp_repo_ubuntu_cloud_archive_version: 20240709T043220
+stackhpc_pulp_repo_ubuntu_jammy_security_version: 20240709T015800
+stackhpc_pulp_repo_ubuntu_jammy_version: 20240709T015800
```
Comment (markgoddard) [inline etc/kayobe/pulp-repo-versions.yml]: I don't think we want to update jammy in this PR.

File: etc/kayobe/pulp-repo-versions.yml [hunk 3]
```diff
@@ -26,12 +26,17 @@ stackhpc_pulp_repo_rocky_9_2_baseos_version: 20231211T120328
...
-stackhpc_pulp_repo_rocky_9_3_appstream_version: 20240416T010957
-stackhpc_pulp_repo_rocky_9_3_baseos_version: 20240413T014042
-stackhpc_pulp_repo_rocky_9_3_crb_version: 20240413T014042
-stackhpc_pulp_repo_rocky_9_3_extras_version: 20240413T014042
-stackhpc_pulp_repo_rocky_9_3_highavailability_version: 20240404T012937
+stackhpc_pulp_repo_rocky_9_3_appstream_version: 20240508T001430
+stackhpc_pulp_repo_rocky_9_3_baseos_version: 20240509T005935
+stackhpc_pulp_repo_rocky_9_3_crb_version: 20240429T003818
+stackhpc_pulp_repo_rocky_9_3_extras_version: 20240506T000343
+stackhpc_pulp_repo_rocky_9_3_highavailability_version: 20240510T001129
```
Comment (markgoddard) [inline etc/kayobe/pulp-repo-versions.yml]: Does the 9.3 update include a kernel bump?

**Review by markgoddard — CHANGES_REQUESTED** [review]

**Comment by markgoddard** [comment]
We do need to consider what to do about the --root-dev-only issue, if it's still a problem.

**Comment by grzegorzkoper** [comment]
> We do need to consider what to do about the --root-dev-only issue, if it's still a problem.

I have not hit this issue.

Issue in SMS-LAB that Piotr was hitting, was that grub.cfg was completely being removed while upgrading packages from 9.3(older kernel) to 9.3 (newer kernel) 
The fix for that was running:
```grub2-mkconfig -o /boot/efi/EFI/rocky/grub.cfg```
After upgrade, before rebooting hosts

File: etc/kayobe/pulp-repo-versions.yml (hunk 3 above)
Comment (grzegorzkoper) [reply etc/kayobe/pulp-repo-versions.yml]: I am not sure. 

The tags were generated automatically by running this workflow:
https://github.com/stackhpc/stackhpc-release-train/actions/runs/9854754622

Based on the instructions here:
https://stackhpc.github.io/stackhpc-release-train/usage/content-workflows/#updating-package-repository-versions-in-kayobe-configuration

Should 9.3 changes be reverted ?

File: etc/kayobe/pulp-repo-versions.yml (hunk 2 above)
Comment (grzegorzkoper) [reply etc/kayobe/pulp-repo-versions.yml]: Thanks, fixed.

File: etc/kayobe/pulp-repo-versions.yml (hunk 3 above)
Comment (mnasiadka) [reply etc/kayobe/pulp-repo-versions.yml]: I don't think we want to touch 9.3

**Comment by markgoddard** [comment]
> > We do need to consider what to do about the --root-dev-only issue, if it's still a problem.
> 
> I have not hit this issue.
> 
> Issue in SMS-LAB that Piotr was hitting, was that grub.cfg was completely being removed while upgrading packages from 9.3(older kernel) to 9.3 (newer kernel) The fix for that was running: `grub2-mkconfig -o /boot/efi/EFI/rocky/grub.cfg` After upgrade, before rebooting hosts

Do you still have an environment running 9.4? If so could you check if `--root-dev-only` is in `/boot/efi/EFI/rocky/grub.cfg`?

**Comment by grzegorzkoper** [comment]
> > > We do need to consider what to do about the --root-dev-only issue, if it's still a problem.
> > 
> > 
> > I have not hit this issue.
> > Issue in SMS-LAB that Piotr was hitting, was that grub.cfg was completely being removed while upgrading packages from 9.3(older kernel) to 9.3 (newer kernel) The fix for that was running: `grub2-mkconfig -o /boot/efi/EFI/rocky/grub.cfg` After upgrade, before rebooting hosts
> 
> Do you still have an environment running 9.4? If so could you check if `--root-dev-only` is in `/boot/efi/EFI/rocky/grub.cfg`?

It's there:
```
[root@gkoper-controller-01 ~]# cat  /boot/efi/EFI/rocky/grub.cfg
search --no-floppy --root-dev-only --fs-uuid --set=dev c209b919-b137-47ad-82ef-bc6cc88323ad
set prefix=($dev)/boot/grub2
export $prefix
configfile $prefix/grub.cfg

[root@gkoper-controller-01 ~]# uname -a
Linux gkoper-controller-01.novalocal 5.14.0-427.22.1.el9_4.x86_64 #1 SMP PREEMPT_DYNAMIC Wed Jun 19 17:35:04 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux
```

grub.cfg on a non upgraded 9.2 machine (control-host)
```[root@gkoper-ansible-control ~]# cat  /boot/efi/EFI/rocky/grub.cfg
#
# DO NOT EDIT THIS FILE
#
# It is automatically generated by grub2-mkconfig using templates
# from /etc/grub.d and settings from /etc/default/grub
#

### BEGIN /etc/grub.d/00_header ###
set pager=1

if [ -f ${config_directory}/grubenv ]; then
  load_env -f ${config_directory}/grubenv
elif [ -s $prefix/grubenv ]; then
  load_env
fi
if [ "${next_entry}" ] ; then
   set default="${next_entry}"
   set next_entry=
   save_env next_entry
   set boot_once=true
else
   set default="0"
fi

if [ x"${feature_menuentry_id}" = xy ]; then
  menuentry_id_option="--id"
else
  menuentry_id_option=""
fi

export menuentry_id_option

if [ "${prev_saved_entry}" ]; then
  set saved_entry="${prev_saved_entry}"
  save_env saved_entry
  set prev_saved_entry=
  save_env prev_saved_entry
  set boot_once=true
fi

function savedefault {
  if [ -z "${boot_once}" ]; then
    saved_entry="${chosen}"
    save_env saved_entry
  fi
}

function load_video {
  if [ x$feature_all_video_module = xy ]; then
    insmod all_video
  else
    insmod efi_gop
    insmod efi_uga
    insmod ieee1275_fb
    insmod vbe
    insmod vga
    insmod video_bochs
    insmod video_cirrus
  fi
}

serial --speed=115200 --unit=0 --word=8 --parity=no --stop=1
terminal_input serial console
terminal_output serial console
if [ x$feature_timeout_style = xy ] ; then
  set timeout_style=menu
  set timeout=5
# Fallback normal timeout code in case the timeout_style feature is
# unavailable.
else
  set timeout=5
fi
### END /etc/grub.d/00_header ###

### BEGIN /etc/grub.d/01_users ###
if [ -f ${prefix}/user.cfg ]; then
  source ${prefix}/user.cfg
  if [ -n "${GRUB2_PASSWORD}" ]; then
    set superusers="root"
    export superusers
    password_pbkdf2 root ${GRUB2_PASSWORD}
  fi
fi
### END /etc/grub.d/01_users ###

### BEGIN /etc/grub.d/08_fallback_counting ###
insmod increment
# Check if boot_counter exists and boot_success=0 to activate this behaviour.
if [ -n "${boot_counter}" -a "${boot_success}" = "0" ]; then
  # if countdown has ended, choose to boot rollback deployment,
  # i.e. default=1 on OSTree-based systems.
  if  [ "${boot_counter}" = "0" -o "${boot_counter}" = "-1" ]; then
    set default=1
    set boot_counter=-1
  # otherwise decrement boot_counter
  else
    decrement boot_counter
  fi
  save_env boot_counter
fi
### END /etc/grub.d/08_fallback_counting ###

### BEGIN /etc/grub.d/10_linux ###
insmod part_gpt
insmod lvm
insmod ext2
set root='lvmid/kdPx7n-cEhR-FwYw-B3cT-G4vT-14fy-Tfocmj/bsOCD3-TKk6-UYY1-3J4j-hfuA-y8g1-ezwNg2'
if [ x$feature_platform_search_hint = xy ]; then
  search --no-floppy --fs-uuid --set=root --hint='lvmid/kdPx7n-cEhR-FwYw-B3cT-G4vT-14fy-Tfocmj/bsOCD3-TKk6-UYY1-3J4j-hfuA-y8g1-ezwNg2'  c209b919-b137-47ad-82ef-bc6cc88323ad
else
  search --no-floppy --fs-uuid --set=root c209b919-b137-47ad-82ef-bc6cc88323ad
fi
insmod part_gpt
insmod lvm
insmod ext2
set boot='lvmid/kdPx7n-cEhR-FwYw-B3cT-G4vT-14fy-Tfocmj/bsOCD3-TKk6-UYY1-3J4j-hfuA-y8g1-ezwNg2'
if [ x$feature_platform_search_hint = xy ]; then
  search --no-floppy --fs-uuid --set=boot --hint='lvmid/kdPx7n-cEhR-FwYw-B3cT-G4vT-14fy-Tfocmj/bsOCD3-TKk6-UYY1-3J4j-hfuA-y8g1-ezwNg2'  c209b919-b137-47ad-82ef-bc6cc88323ad
else
  search --no-floppy --fs-uuid --set=boot c209b919-b137-47ad-82ef-bc6cc88323ad
fi

# This section was generated by a script. Do not modify the generated file - all changes
# will be lost the next time file is regenerated. Instead edit the BootLoaderSpec files.
#
# The blscfg command parses the BootLoaderSpec files stored in /boot/loader/entries and
# populates the boot menu. Please refer to the Boot Loader Specification documentation
# for the files format: https://systemd.io/BOOT_LOADER_SPECIFICATION/.

# The kernelopts variable should be defined in the grubenv file. But to ensure that menu
# entries populated from BootLoaderSpec files that use this variable work correctly even
# without a grubenv file, define a fallback kernelopts variable if this has not been set.
#
# The kernelopts variable in the grubenv file can be modified using the grubby tool or by
# executing the grub2-mkconfig tool. For the latter, the values of the GRUB_CMDLINE_LINUX
# and GRUB_CMDLINE_LINUX_DEFAULT options from /etc/default/grub file are used to set both
# the kernelopts variable in the grubenv file and the fallback kernelopts variable.
if [ -z "${kernelopts}" ]; then
  set kernelopts="root=LABEL=rootfs ro  console=tty0 console=ttyS0,115200 no_timer_check nofb nomodeset gfxpayload=text net.ifnames=1 rd.auto  "
fi

insmod blscfg
blscfg
### END /etc/grub.d/10_linux ###

### BEGIN /etc/grub.d/10_reset_boot_success ###
# Hiding the menu is ok if last boot was ok or if this is a first boot attempt to boot the entry
if [ "${boot_success}" = "1" -o "${boot_indeterminate}" = "1" ]; then
  set menu_hide_ok=1
else
  set menu_hide_ok=0
fi
# Reset boot_indeterminate after a successful boot
if [ "${boot_success}" = "1" ] ; then
  set boot_indeterminate=0
# Avoid boot_indeterminate causing the menu to be hidden more than once
elif [ "${boot_indeterminate}" = "1" ]; then
  set boot_indeterminate=2
fi
# Reset boot_success for current boot
set boot_success=0
save_env boot_success boot_indeterminate
### END /etc/grub.d/10_reset_boot_success ###

### BEGIN /etc/grub.d/12_menu_auto_hide ###
### END /etc/grub.d/12_menu_auto_hide ###

### BEGIN /etc/grub.d/14_menu_show_once ###
if [ x$feature_timeout_style = xy ]; then
  if [ "${menu_show_once_timeout}" ]; then
    set timeout_style=menu
    set timeout="${menu_show_once_timeout}"
    unset menu_show_once_timeout
    save_env menu_show_once_timeout
  fi
fi
### END /etc/grub.d/14_menu_show_once ###

### BEGIN /etc/grub.d/20_linux_xen ###
### END /etc/grub.d/20_linux_xen ###

### BEGIN /etc/grub.d/20_ppc_terminfo ###
### END /etc/grub.d/20_ppc_terminfo ###

### BEGIN /etc/grub.d/30_os-prober ###
### END /etc/grub.d/30_os-prober ###

### BEGIN /etc/grub.d/30_uefi-firmware ###
if [ "$grub_platform" = "efi" ]; then
	menuentry 'UEFI Firmware Settings' $menuentry_id_option 'uefi-firmware' {
		fwsetup
	}
fi
### END /etc/grub.d/30_uefi-firmware ###

### BEGIN /etc/grub.d/35_fwupd ###
### END /etc/grub.d/35_fwupd ###

### BEGIN /etc/grub.d/40_custom ###
# This file provides an easy way to add custom menu entries.  Simply type the
# menu entries you want to add after this comment.  Be careful not to change
# the 'exec tail' line above.
### END /etc/grub.d/40_custom ###

### BEGIN /etc/grub.d/41_custom ###
if [ -f  ${config_directory}/custom.cfg ]; then
  source ${config_directory}/custom.cfg
elif [ -z "${config_directory}" -a -f  $prefix/custom.cfg ]; then
  source $prefix/custom.cfg
fi
### END /etc/grub.d/41_custom ###
[root@gkoper-ansible-control ~]# uname -a
Linux gkoper-ansible-control.novalocal 5.14.0-284.30.1.el9_2.x86_64 #1 SMP PREEMPT_DYNAMIC Sat Sep 16 09:55:41 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux
```

File: etc/kayobe/pulp-repo-versions.yml (hunk 3 above)
Comment (markgoddard) [reply etc/kayobe/pulp-repo-versions.yml]: Done

**Review by markgoddard — APPROVED** [review]
I'm going to stop blocking this based on the --root-dev-only issue because it also affects our 9.3 snapshots. We should still try to address it though.

**Pull request merged**